flow.py

```diff
@@ -57,17 +67,79 @@ def dest(self):
     def dest(self, dest):
         raise AttributError("Cannot modify a flow's destination node.")
 
-    def make_packet(self, id, src, dest, data):
+    def check_last(self):
+        """Method that checks if all packets have been acknowledged and
+        updating the NetworkSimulator object accordingly
+        """
+        if self.last_acknowledged == self.num_packets:
+            self.ns.num_active_flows -= 1
+
+    def update_flow(a_packet):
+        """Upon receiving an acknowledgement packet, updating the flow's
+        attributes
+
+        Args:
+            a_packet (AcknowledgementPacket): Packet being sent
+                back from host
+        """
+        print("Flow: updated with acknowledgement packet: ", a_packet.id,
+            "And flow", self.flow_id)
+        self.check_last()
+        self.unacknowledged_packets.remove(a_packet.packet_id)
+
+
+    def make_data_packet(self, src, dest, packet_size, data):
         """Method makes the packet and triggers the send_packet method for the
         host if applicable
 
         Args:
-            id (int): The unique id identifying the packet
+            packet_id (string): Unique id identifying the packet
             src (Node): The packet's source node
             dest (Node): The packet's destination node
+            packet_size (float): The packet's size in bits
+            flow_id (string): Unique id indicating packet
             data (any datatype): The data in the packet
         """
-        pass
+        print("Flow: made data packet", self.flow_id)
+        new_packet = DataPacket(ns, self.num_packets_sent + 1, src, dest,
+            self.flow_id, packet_size, data)
+
+        event = lambda: self.ns.hosts[new_packet.src].send_packet(new_packet)
+        self.ns.add_event(event)
+
+        self.num_packets_sent += 1
+
+    def make_acknowledgment_packet(self, packet_id, src, dest, packet_size, acknowledge_id):
+        """Method makes the packet and triggers the send_packet method for the
+        host if applicable
+
+        Args:
+            packet_id (string): Unique id identifying the packet
+            src (Node): The packet's source node
+            dest (Node): The packet's destination node
+            packet_size (float): The packet's size in bits
+            flow_id (string): Unique id indicating packet
+            acknowledge_id (string): Unique id identifying packet being acknowledged
+        """
+        print("Flow: made acknowledgement packet", packet.packet_id,
+            "in flow", flow.flow_id)
```

Comment:
shouldn't you print this after you actually make the packet? also packet.packet_id isn't a thing here